src/pykeen/nn/init.py

```diff
@@ -308,6 +311,84 @@ def from_triples_factory(
+
+    .. math::
+        \mathbf{x}_i = [\mathbf{R}_{i, i}, \mathbf{R}^{2}_{i, i}, \ldots, \mathbf{R}^{d}_{i, i}] \in \mathbb{R}^{d}
+
+    where $\mathbf{R} := \mathbf{A}\mathbf{D}^{-1}$ is the random walk matrix, with
+    $\mathbf{D} := \sum_i \mathbf{A}_{i, i}$.
+
+    .. seealso::
+        https://arxiv.org/abs/2110.07875
+    """
+
+    def __init__(
+        self,
+        *,
+        triples_factory: Optional[CoreTriplesFactory] = None,
+        mapped_triples: Optional[MappedTriples] = None,
+        edge_index: Optional[torch.Tensor] = None,
+        dim: int,
+        num_entities: Optional[int] = None,
+        space_dim: Optional[int] = 0,
+    ) -> None:
+        """
+        Initialize the positional encoding.
+
+        One of `triples_factory`, `mapped_triples` or `edge_index` will be used.
+        The preference order is:
+
+        1. `triples_factory`
+        2. `mapped_triples`
+        3. `edge_index`
+
+        :param triples_factory:
+            the triples factory
+        :param mapped_triples: shape: `(m, 3)`
+            the mapped triples
+        :param edge_index: shape: `(2, m)`
+            the edge index
+        :param dim:
+            the dimensionality
+        :param num_entities:
+            the number of entities. If `None`, it will be inferred from `edge_index`
+        :param space_dim:
+            estimated dimensionality of the space. Used to
+            correct the random-walk diagonal by a factor `k^(space_dim/2)`.
+            In euclidean space, this correction means that the height of
+            the gaussian distribution stays almost constant across the number of
+            steps, if `space_dim` is the dimension of the euclidean space.
+        """
+        if triples_factory is not None:
+            mapped_triples = triples_factory.mapped_triples
+        if mapped_triples is not None:
+            edge_index = mapped_triples[:, 0::2].t()
+        # create random walk matrix
+        num_entities = torch_ppr.utils.prepare_num_nodes(edge_index=edge_index, num_nodes=num_entities)
+        adj = torch_ppr.utils.edge_index_to_sparse_matrix(edge_index=edge_index, num_nodes=num_entities)
+        adj = torch_ppr.utils.prepare_page_rank_adjacency(edge_index=edge_index)
+        # allocate tensor
+        tensor = torch.empty(num_entities, dim)
+        # note: create tensor to extract diagonals
+        # torch.diag does not work with sparse tensors
+        diag_indices = torch.arange(num_entities).unsqueeze(0).repeat(2, 1)
+        one_diag = torch.sparse_coo_tensor(indices=diag_indices, values=torch.ones(num_entities))
+        # iterate
+        for i in range(dim):
+            adj = torch.sparse.mm(adj, adj)
+            # extract diagonal; torch.diag does not work with sparse tensors
+            diag = (adj * one_diag).coalesce()
+            indices = diag.indices()
+            assert (indices == diag_indices).all()
+            tensor[:, i] = diag.values() * (i ** (space_dim / 2))
```

Comment:
I added an optional diffusion normalization parameter (defaults to do nothing) - we found it is useful from our experiments

Reply:
are there any precision / complexity issues that could arise from doing this `i**0` operation and multiplying? would it be worth using an if statement to check if space_dim is set and do it otherwise not?

Reply:
yep, fixed that in https://github.com/pykeen/pykeen/pull/918/commits/f6a39685e65c35ba6a3d7450c6c85e0c9b8a6bd9